### PR TITLE
10186: Re-fix cancellation of deferred flattening

### PR DESCRIPTION
**Pull request by mthuurne**
## Scope and purpose

The previous fix (#1590) was incorrect and even though it included a test case, the test wasn't sensitive enough to detect the problem.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10186
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green. -> failing test is a known issue unrelated to this PR
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: mthuurne
Reviewer: glyph
Fixes: ticket:10186

Properly fix the cancellation of deferred flattening and improve the test case.
```

### CODE REVIEW

File: src/twisted/web/test/test_flatten.py [hunk 1]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
         """
         The flattening of a Deferred can be cancelled.
         """
-        d = Deferred()
+        cancel_count = 0
+        cancel_arg = None
+
+        def check_cancel(cancelled):
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
        def checkCancel(cancelled):
```
Let's stick to Twisted house style please ;)

File: src/twisted/web/test/test_flatten.py [hunk 2]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
...
+            nonlocal cancel_arg, cancel_count
+            cancel_arg = cancelled
+            cancel_count += 1
+
+        err = None
+
+        def save_err(failure):
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
        def saveErr(failure):
```

File: src/twisted/web/test/test_flatten.py [hunk 3]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
         """
         The flattening of a Deferred can be cancelled.
         """
-        d = Deferred()
+        cancel_count = 0
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
        cancelCount = 0
```

File: src/twisted/web/test/test_flatten.py [hunk 4]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
...
+        cancel_arg = None
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
        cancelArg = None
```

File: src/twisted/web/test/test_flatten.py [hunk 5]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
...
+
+        def check_cancel(cancelled):
+            nonlocal cancel_arg, cancel_count
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
            nonlocal cancelArg, cancelCount
```

File: src/twisted/web/test/test_flatten.py [hunk 6]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
...
+            cancel_arg = cancelled
+            cancel_count += 1
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
            cancelArg = cancelled
            cancelCount += 1
```

File: src/twisted/web/test/test_flatten.py [hunk 7]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
...
+
+        err = None
+
+        def save_err(failure):
+            nonlocal err
+            err = failure
+
+        d = Deferred(check_cancel)
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
        d = Deferred(checkCancel)
```

File: src/twisted/web/test/test_flatten.py [hunk 8]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
...
         flattening = flattenString(None, d)
         self.assertNoResult(flattening)
+        d.addErrback(save_err)
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
        d.addErrback(saveErr)
```

File: src/twisted/web/test/test_flatten.py [hunk 9]
```diff
@@ -574,12 +575,36 @@ def test_cancel(self):
...
 
         flattening.cancel()
 
+        # Check whether we got an orderly cancellation.
+        # Do this first to get more meaningful reporting if something crashed.
         failure = self.failureResultOf(flattening, FlattenerError)
+
+        self.assertEqual(cancel_count, 1)
+        self.assertIs(cancel_arg, d)
```
Comment (glyph) [inline src/twisted/web/test/test_flatten.py]:
```suggestion
        self.assertEqual(cancelCount, 1)
        self.assertIs(cancelArg, d)
```

**Review by glyph — APPROVED** [review]
Aside from style nitpicks, LGTM. Do you mind addressing these?

**Comment by mthuurne** [comment]
> Aside from style nitpicks, LGTM. Do you mind addressing these?

No problem; I used lower case without thinking because most projects I'm currently working on use it.

**Pull request merged**